crypto/hpke/hpke.c

```diff
@@ -0,0 +1,2366 @@
+    if (ikm != NULL) {
+        *p++ = OSSL_PARAM_construct_octet_string(OSSL_PKEY_PARAM_DHKEM_IKM,
+                                                 (char *)ikm, ikmlen);
+    }
+    *p = OSSL_PARAM_construct_end();
+    if (EVP_PKEY_CTX_set_params(pctx, params) <= 0) {
+        erv = 0;
+        ERR_raise(ERR_LIB_CRYPTO, ERR_R_INTERNAL_ERROR);
+        goto err;
+    }
+    if (EVP_PKEY_generate(pctx, &skR) <=0) {
+        erv = 0;
+        ERR_raise(ERR_LIB_CRYPTO, ERR_R_INTERNAL_ERROR);
+        goto err;
+    }
+    EVP_PKEY_CTX_free(pctx);
+    pctx = NULL;
+    lpublen = EVP_PKEY_get1_encoded_public_key(skR, &lpub);
+    if (lpub == NULL || lpublen == 0) {
+        erv = 0;
+        ERR_raise(ERR_LIB_CRYPTO, ERR_R_INTERNAL_ERROR);
+        goto err;
+    }
+    if (lpublen > *publen) {
+        erv = 0;
+        ERR_raise(ERR_LIB_CRYPTO, ERR_R_INTERNAL_ERROR);
+        goto err;
+    }
+    *publen = lpublen;
+    memcpy(pub, lpub, lpublen);
+    *priv = skR;
+
+err:
+    if (erv != 1) { EVP_PKEY_free(skR); }
+    EVP_PKEY_CTX_free(pctx);
+    OPENSSL_free(lpub);
+    return erv;
+}
+
+/*
+ * @brief randomly pick a suite
+ *
+ * @param libctx is the context to use
+ * @param propq is a properties string
+ * @param suite is the result
+ * @return 1 for success, otherwise failure
+ *
+ * If you change the structure of the various *_tab arrays
+ * then this code will also need change.
+ */
+static int hpke_random_suite(OSSL_LIB_CTX *libctx,
+                             const char *propq,
+                             OSSL_HPKE_SUITE *suite)
+{
+    unsigned char rval = 0;
+    int nkdfs = OSSL_NELEM(hpke_kdf_tab)-1;
+    int naeads = OSSL_NELEM(hpke_aead_tab)-1;
+    int nkems = OSSL_NELEM(hpke_kem_tab)-1;
+
+    /* random kem */
+    if (RAND_bytes_ex(libctx, &rval, sizeof(rval), OSSL_HPKE_RSTRENGTH) <= 0)
+        return 0;
+    suite->kem_id = hpke_kem_tab[(rval % nkems + 1)].kem_id;
+
+    /* random kdf */
+    if (RAND_bytes_ex(libctx, &rval, sizeof(rval), OSSL_HPKE_RSTRENGTH) <= 0)
+        return 0;
+    suite->kdf_id = hpke_kdf_tab[(rval % nkdfs + 1)].kdf_id;
+
+    /* random aead */
+    if (RAND_bytes_ex(libctx, &rval, sizeof(rval), OSSL_HPKE_RSTRENGTH) <= 0)
+        return 0;
+    suite->aead_id = hpke_aead_tab[(rval % naeads + 1)].aead_id;
+    return 1;
+}
+
+/*
+ * @brief return a (possibly) random suite, public key, ciphertext for GREASErs
+ *
+ * @param libctx is the context to use
+ * @param propq is a properties string
+ * @param suite-in specifies the preferred suite or NULL for a random choice
+ * @param suite is the chosen or random suite
+ * @param pub a random value of the appropriate length for sender public value
+ * @param pub_len is the length of pub (buffer size on input)
+ * @param ciphertext buffer with random value of the appropriate length
+ * @param ciphertext_len is the length of cipher
+ * @return 1 for success, otherwise failure
+ */
+static int hpke_good4grease(OSSL_LIB_CTX *libctx, const char *propq,
```

Comment:
hpke_get_grease_values() maybe ??

Reply:
that's better yes, will change, maybe singular rather than plural though